cmdx.py

```diff
@@ -3861,13 +3948,144 @@ def _python_to_mod(value, plug, mod):
     return True
 
 
+class MetaComponent(type):
+
+    @withTiming()
+    def __call__(cls, mdagpath, mobject, exists=True, modifier=None):
+
+        if mobject.hasFn(om.MFn.kMeshVertComponent):
+            sup = MeshVertex
+        else:
+            sup = Component
+
+        self = super(MetaComponent, sup).__call__(mdagpath, mobject, exists, modifier)
+        return self
+
+
+@add_metaclass(MetaComponent)
+class Component(object):
+    _Fn = om.MFnComponent
+
+    typeName = None
+
+    def __init__(self, node, mobject, exists=True, modifier=None):
+        """A Maya component container
+
+        Arguments:
+            node (DagNode): cmdx shape instance
+            mobject (maya.api.OpenMaya.MObject): Internal component object
+        """
+        assert isinstance(node, DagNode), "%s is not a DagNode" % node
+        assert isinstance(mobject, om.MObject), "%s is not a MObject" % mobject
+
+        self._node = node
+        self._mdagpath = node.dagPath()
+
+        self._mobject = mobject
+        self._fn = self._Fn(mobject)
+
+        self._it = None
+
+    def __str__(self):
+        """Return component selection string
+        """
+        sl = om.MSelectionList()
+        sl.add((self._mdagpath, self._mobject))
+        return ' '.join(sl.getSelectionStrings(0))
+
+    path = __str__
+
+    def __nonzero__(self):
+        return not self._fn.isEmpty
+
+    __bool__ = __nonzero__
+
+    @property
+    def complete(self):
+        return self._fn.isComplete
+
+
+class Component1D(Component):
+    _Fn = om.MFnSingleIndexedComponent
+
+    def __len__(self):
+        return self._fn.getCompleteData()
+
+    def __contains__(self, item):
+        return item in self._fn.getElements()
+
+    def __iter__(self):
+        """Iterate over the component indices
+        """
+        if self._fn.isComplete:
+            for i in xrange(len(self)):
+                yield i
+        else:
+            for i in self._fn.getElements():
+                yield i
+
+    def __getitem__(self, index):
+        cps = self._Fn().create(self._type)
+        fn = self._Fn(cps)
+        n = self.maxLength()
+        if n is None:
+            raise IndexError("{} doesn't support get item".format(self.__class__))
+
+        if isinstance(index, slice):
+            fn.addElements(range(*index.indices(n)))
+        elif isinstance(index, tuple):
+            fn.addElements(index)
+        else:
+            if index < 0:
+                index = n + 1 + index
+            fn.addElement(index)
+
+        return self.__class__(self._node, cps)
```

Comment:
Can we use `type(self)` instead of `__class__`? It's used elsewhere for this purpose.